workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceFileSystemService.java

```
@@ -0,0 +1,312 @@
+
+
+      // Walk source directory and copy files/subdirectories -- note we have to use a try-with-resources thing here
+      // to ensure the stream autocloses
+      try (var paths = Files.walk(sourcePath)) {
+        paths.forEach(source -> {
+          final Path relative = sourcePath.relativize(source);
+          final Path target = destPath.resolve(relative);
+          try {
+            if (Files.isDirectory(source)) {
+              Files.createDirectories(target);
+            } else {
+              Files.copy(source, target);
+            }
+          } catch (IOException e) {
+            throw new UncheckedIOException(e);
+          }
+        });
+      }
+
+      return true;
+    } catch (IOException | UncheckedIOException e) {
+      logger.error("Error copying directory", e);
+      return false;
+    }
+  }
+
+
+
+  @Override
+  public boolean deleteFile(final int workspaceId, final Path filePath) throws NoSuchWorkspaceException {
+    final var repoPath = postgresRepository.workspaceRootPath(workspaceId);
+    final var file = repoPath.resolve(filePath).toFile();
+    return file.delete();
+  }
+
+  @Override
+  public DirectoryTree listFiles(final int workspaceId, final Optional<Path> directoryPath, final int depth)
+  throws SQLException, NoSuchWorkspaceException {
+    final var repoPath = postgresRepository.workspaceRootPath(workspaceId);
+    final var path = repoPath.resolve(directoryPath.orElse(Path.of("")));
+
+    if(!path.toFile().isDirectory()) {
+      return null;
+    }
+
+    // Converting our API to the Files API
+    final var walkDepth = depth == -1 ? Integer.MAX_VALUE : depth + 1;
+    try(final Stream<Path> walkOutput = Files.walk(path, walkDepth)) {
+      final var walkList = new ArrayList<>(walkOutput.toList());
+      walkList.removeFirst(); // remove the initial path
+      return new DirectoryTree(path, walkList, postgresRepository.getExtensionMapping());
+    } catch (IOException io) {
+      return null;
```

We should probably throw this Exception up to the caller to get the message - as-is any IOException will be reported as "directory does not exist" but there could be other reasons for the failure

---

Exception is now bubbled-up